src/java.base/share/classes/java/lang/String.java

```diff
@@ -2987,7 +2987,10 @@ public String concat(String str) {
         if (str.isEmpty()) {
             return this;
         }
-        return StringConcatHelper.simpleConcat(this, str);
+        if (isEmpty()) {
+            return str;
```

Comment:
This case should probably be reflected more precisely in the specification, or do `return new String(str);` to avoid changing semantics for this corner-case.

Reply:
Yes, this isn't beholden to JLS 15.18.1, and it's already specified that `foo.concat("")` returns `foo` - so why shouldn't `"".concat(foo)` return `foo`? But still it's an observable semantic change so some care needs to be taken - possibly even a CSR is warranted. Doing `return new String(str)` to retain behavior avoids that headache for a fringe case.

Reply:
Adding isEmpty check is not related to this PR and should be a new PR.

Reply:
Thanks. Removing this check will make the `"".append(foo)` edge case a little bit slower (likely inconsequential), but removes a branch and might thus even be an optimization in the common case.